### ISSUE TITLE
Add support for Argon2 password hashing

### ISSUE THREAD

**Issue by mmd-osm** [zerebubuth/openstreetmap-cgimap#250]
See https://github.com/openstreetmap/openstreetmap-website/pull/3353

TODOs:
* [x] Add argon2 verify_password in https://github.com/zerebubuth/openstreetmap-cgimap/blob/master/src/basicauth.cpp
* [x] Add test cases: https://github.com/zerebubuth/openstreetmap-cgimap/blob/master/test/test_basicauth.cpp
* [x] Update https://github.com/zerebubuth/openstreetmap-cgimap/blob/master/Dockerfile
* [x] Add  libargon2-dev https://github.com/zerebubuth/openstreetmap-cgimap/blob/ubuntu/focal/debian/control
* [x] Update lib install instructions https://github.com/zerebubuth/openstreetmap-cgimap/blob/master/README#L55
* [x] Update configure.ac w/ libargon2
* [x] End-to-end test Rails port -> cgimap
* [x] Deploy libargon2-dev on errol
* [x] Add libargon2-1 in Dockerfile
* [x] New release 0.8.6
* [x] Building new PPA, https://launchpad.net/~mmd-osm/+archive/ubuntu/cgimap-806-focal
* [x] Test PPA
<details>

Test package on ubuntu 20.04 container: deps ok.
```
 openstreetmap-cgimap --help
CGImap 0.8.6: Allowed options:
  --help                        display this help and exit
  --daemon                      run as a daemon
  --instances arg (=5)          number of daemon instances to run
  --pidfile arg                 file to write pid to
```
</details>

* [x] Copy PPA over to https://launchpad.net/~mmd-osm/+archive/ubuntu/cgimap

**Comment by mmd-osm** [comment]
I don't think `argon2_verify` will work when using a secret (PEPPER). The function interface doesn't expose a parameter for the secret. OTOH `argon2_verify_ctx ` depends on decode_string() down the road which isn't exposed by the library and needs to be reimplemented (according to some issue I've found on the argon2 issue tracker). Probably the Ruby library did something similar in https://github.com/technion/ruby-argon2/blob/master/ext/argon2_wrap/argon_wrap.c#L101-L167

I'm not super excited about those workarounds and additional dependencies, and it might take a while to sort this all out.

**Comment by mmd-osm** [comment]
So the Ruby gem is using a local copy of the argon2 C library in https://github.com/technion/ruby-argon2/tree/master/ext to include the "decode_string" function straight from the source code. The Ubuntu package you've mentioned only exposes argon2.h, which is the official API (see https://packages.debian.org/sid/amd64/libargon2-dev/filelist). Directly including any security related source code raises a few red flags here.

Regarding the type, I think the expectation is to parse the "$argo2..:" hash value and figure out for yourself what the respective type is. Again, this is a big no go. I don't want to be concerned with implementation level details of the argon2 library. To me, the hash value is an opaque string. All I want is to provide the hash and a password to some function and have the library figure out the rest, obviously using the official API.

https://github.com/technion/ruby-argon2/blob/master/ext/argon2_wrap/argon_wrap.c#L125-L136

**Comment by tzapzoor** [comment]
Hey @mmd-osm. Was playing around with this project's Dockerfile in the master branch and noticed that you also need to install `libargon2-0` via `apt` otherwise you get this error `libargon2.so.0: cannot open shared object file: No such file or directory`.

**Comment by mmd-osm** [comment]
Yes, that's spot on, will do.

**Comment by mmd-osm** [comment]
I added libargon2-1  instead, seems to work ok here.

**Comment by mmd-osm** [comment]
Fix pushed as e1dd56e5d6775f4f82876ff87c730b57bf5b9855
